File: packages/core/src/reducer.ts
import { GameState, GameAction, GameMode, GameStatus, BoardState, GauntletStageConfig, GauntletStageResult, GAUNTLET_STAGES, GAUNTLET_TOTAL_SOLUTIONS } from './types';
import { generateSolutionsFromSeed } from './seed';
import { evaluateGuess } from './evaluator';
import { isValidWord, getAllowedWords } from './dictionary';
import { generatePrefillGuesses, generatePrefillWords } from './prefill';

export function initializeGame(seed: string, mode: GameMode): GameState {
  return createInitialState(seed, mode);
}

function createBoardState(solution: string, maxGuesses: number = 6): BoardState {
  return {
    solution,
    guesses: [],
    maxGuesses,
    status: GameStatus.PLAYING
  };
}

function createStageBoardsFromSolutions(
  seed: string,
  stage: GauntletStageConfig,
  solutions: string[]
): BoardState[] {
  const boards = solutions.map(sol => createBoardState(sol, stage.maxGuesses));

  if (stage.hasPrefill) {
    const allowedWords = getAllowedWords();
    const prefillWords = generatePrefillWords(seed, solutions, allowedWords);
    boards.forEach((board) => {
      board.prefilledGuesses = generatePrefillGuesses(prefillWords, board.solution);
    });
  }

  return boards;
}

function getStageSolutionSlice(allSolutions: string[], stageIndex: number): string[] {
  let offset = 0;
  for (let i = 0; i < stageIndex; i++) {
    offset += GAUNTLET_STAGES[i].boardCount;
  }
  return allSolutions.slice(offset, offset + GAUNTLET_STAGES[stageIndex].boardCount);
}

export function createInitialState(seed: string, mode: GameMode): GameState {
  let boardCount = 1;
  let maxGuesses = 6;

  switch (mode) {
    case GameMode.DUEL:
      boardCount = 1;
      maxGuesses = 6;
      break;
    case GameMode.MULTI_DUEL:
      boardCount = 2;
      maxGuesses = 6;
      break;
    case GameMode.GAUNTLET: {
      const allSolutions = generateSolutionsFromSeed(seed, GAUNTLET_TOTAL_SOLUTIONS);
      const firstStage = GAUNTLET_STAGES[0];
      const firstStageSolutions = getStageSolutionSlice(allSolutions, 0);
      const gauntletBoards = createStageBoardsFromSolutions(seed, firstStage, firstStageSolutions);

      return {
        mode,
        seed,
        startTime: Date.now(),
        boards: gauntletBoards,
        currentBoardIndex: 0,
        status: GameStatus.PLAYING,
        gauntlet: {
          currentStage: 0,
          totalStages: GAUNTLET_STAGES.length,
          stages: GAUNTLET_STAGES,
          stageResults: [],
          stageStartTime: Date.now(),
          stageStartElapsedMs: 0,
          allSolutions,
          blackoutCount: 0
        }
      };
    }
    case GameMode.QUORDLE:
      boardCount = 4;
      maxGuesses = 9;
      break;
    case GameMode.OCTORDLE:
      boardCount = 8;
      maxGuesses = 13;
      break;
    case GameMode.SEQUENCE:
      boardCount = 4;
      maxGuesses = 10;
      break;
    case GameMode.RESCUE:
      boardCount = 4;
      maxGuesses = 6;
      break;
    case GameMode.TOURNAMENT:
      boardCount = 5;
      maxGuesses = 6;
      break;
  }

  const solutions = generateSolutionsFromSeed(seed, boardCount);
  const boards = solutions.map(sol => createBoardState(sol, maxGuesses));

  if (mode === GameMode.RESCUE) {
    const allowedWords = getAllowedWords();
    const prefillWords = generatePrefillWords(seed, solutions, allowedWords);
    boards.forEach((board) => {
      board.prefilledGuesses = generatePrefillGuesses(prefillWords, board.solution);
    });
  }

  const state: GameState = {
    mode,
    seed,
    startTime: Date.now(),
    boards,
    currentBoardIndex: 0,
    status: GameStatus.PLAYING
  };

  return state;
}

export function gameReducer(state: GameState, action: GameAction): GameState {
  switch (action.type) {
    case 'SUBMIT_GUESS': {
      const { guess, boardIndex = state.currentBoardIndex } = action;

      // NOTE: we deliberately do NOT gate on state.status here. Multi-
      // board modes (Rescue, Quordle, Octordle, Sequence, Gauntlet
      // stages) submit the same user guess as one dispatch per playing
      // board. A previous iteration of this reducer rejected any guess
      // whose first sibling-dispatch had already tipped the game into
      // LOST (e.g. the losing-guess also happened to be the answer on
      // another board — it got dropped). The per-board check below is
      // the authoritative gate; the game-status update at the bottom
      // still winds up correct either way.

      if (guess.length !== 5) {
        return state;
      }

      if (!isValidWord(guess)) {
        return state;
      }

      const board = state.boards[boardIndex];
      if (board.status !== GameStatus.PLAYING) {
        return state;
      }

      const result = evaluateGuess(board.solution, guess);
      const newGuesses = [...board.guesses, guess.toUpperCase()];

      let newStatus: GameStatus = board.status;
      if (result.isCorrect) {
        newStatus = GameStatus.WON;
      } else if (newGuesses.length >= board.maxGuesses) {
        newStatus = GameStatus.LOST;
      }

      const newBoards = [...state.boards];
      newBoards[boardIndex] = {
        ...board,
        guesses: newGuesses,
        status: newStatus
      };

      let gameStatus: GameStatus = state.status;

      if (state.mode === GameMode.DUEL) {
        gameStatus = newStatus;
      } else if (state.mode === GameMode.MULTI_DUEL) {
        const allComplete = newBoards.every(b => b.status !== GameStatus.PLAYING);
        if (allComplete) {
          const anyWon = newBoards.some(b => b.status === GameStatus.WON);
          gameStatus = anyWon ? GameStatus.WON : GameStatus.LOST;
        }
      } else if (state.mode === GameMode.GAUNTLET && state.gauntlet) {
        // A failed board ends the run immediately — same semantics as
        // every other solo mode: you ran out of guesses on a stage, you
        // are done. The previous version of this branch kept the game
        // PLAYING and let the UI trigger a Letter Blackout second-chance
        // mechanic, but that was a VS-style catch-up that didn't belong
        // in single-player and was reported as a bug (losing OctoWord
        // in the daily gauntlet would silently restart the failed board
        // instead of ending the run). VS Gauntlet still routes loss
        // accounting through state.status === LOST downstream, so it
        // picks up the new timing without any VS-specific wiring.
        const anyLost = newBoards.some(b => b.status === GameStatus.LOST);
        if (anyLost) {
          // Push the failed stage into stageResults so GauntletResults
          // renders a red "failed here" row for the stage that killed
          // the run. NEXT_STAGE only appends on successful completion,
          // so without this the failed stage would be invisible in the
          // per-stage breakdown.
          const stageGuesses = newBoards.reduce((max, b) => Math.max(max, b.guesses.length), 0);
          const stageTimeMs = Date.now() - state.gauntlet.stageStartTime;
          const failedStageResult: GauntletStageResult = {
            stageIndex: state.gauntlet.currentStage,
            status: GameStatus.LOST,
            guesses: stageGuesses,
            timeMs: stageTimeMs,
          };
          return {
            ...state,
            boards: newBoards,
            status: GameStatus.LOST,
            gauntlet: {
              ...state.gauntlet,
              stageResults: [...state.gauntlet.stageResults, failedStageResult],
            },
          };
        }
      } else if (state.mode === GameMode.QUORDLE || state.mode === GameMode.OCTORDLE || state.mode === GameMode.SEQUENCE) {
        // All these modes: every guess goes to all boards, win when all solved, lose when out of guesses
        const allComplete = newBoards.every(b => b.status !== GameStatus.PLAYING);
        if (allComplete) {
          const allWon = newBoards.every(b => b.status === GameStatus.WON);
          gameStatus = allWon ? GameStatus.WON : GameStatus.LOST;
        }
      } else if (state.mode === GameMode.RESCUE || state.mode === GameMode.TOURNAMENT) {
        if (newStatus === GameStatus.LOST) {
          gameStatus = GameStatus.LOST;
        } else if (newStatus === GameStatus.WON) {
          const allWon = newBoards.every(b => b.status === GameStatus.WON);
          if (allWon) {
            gameStatus = GameStatus.WON;
          }
        }
      }

      return {
        ...state,
        boards: newBoards,
        status: gameStatus
      };
    }

    case 'NEXT_BOARD': {
      // For sequential modes: advance to the next board within the current stage/game
      const currentBoard = state.boards[state.currentBoardIndex];
      if (currentBoard.status !== GameStatus.WON) {
        return state;
      }

      const nextIndex = state.currentBoardIndex + 1;
      if (nextIndex >= state.boards.length) {
        return state;
      }

      return {
        ...state,
        currentBoardIndex: nextIndex
      };
    }

    case 'NEXT_STAGE': {
      if (state.mode !== GameMode.GAUNTLET || !state.gauntlet) {
        return state;
      }

      const gauntlet = state.gauntlet;
      const currentStageConfig = gauntlet.stages[gauntlet.currentStage];

      // Verify current stage is complete (all boards won)
      const stageComplete = state.boards.every(b => b.status === GameStatus.WON);
      if (!stageComplete) {
        return state;
      }

      // Stage duration preference order:
      // 1. elapsedMs from the caller (active-play clock from the solo
      //    gauntlet's useActivePlayTimer) minus stageStartElapsedMs. This
      //    excludes any time the tab was hidden.
      // 2. Wall-clock Date.now() - stageStartTime. VS gauntlet uses this
      //    path so the opponent's and our clocks agree.
      const stageTimeMs =
        action.elapsedMs !== undefined && gauntlet.stageStartElapsedMs !== undefined
          ? Math.max(0, action.elapsedMs - gauntlet.stageStartElapsedMs)
          : Date.now() - gauntlet.stageStartTime;

      // Record stage result.
      // Every guess is dispatched to all PLAYING boards simultaneously, so the
      // board that took the longest to solve always holds every guess typed
      // during this stage. Using sum would multi-count each guess per board.
      const stageResult = {
        stageIndex: gauntlet.currentStage,
        status: GameStatus.WON,
        guesses: state.boards.reduce((max, b) => Math.max(max, b.guesses.length), 0),
        timeMs: stageTimeMs
      };

      const newStageResults = [...gauntlet.stageResults, stageResult];
      const nextStageIndex = gauntlet.currentStage + 1;

      // If that was the last stage, the whole gauntlet is won
      if (nextStageIndex >= gauntlet.totalStages) {
        return {
          ...state,
          status: GameStatus.WON,
          gauntlet: {
            ...gauntlet,
            stageResults: newStageResults
          }
        };
      }

      // Load next stage
      const nextStage = gauntlet.stages[nextStageIndex];
      const nextSolutions = getStageSolutionSlice(gauntlet.allSolutions, nextStageIndex);
      const nextBoards = createStageBoardsFromSolutions(state.seed, nextStage, nextSolutions);

      return {
        ...state,
        boards: nextBoards,
        currentBoardIndex: 0,
        gauntlet: {
          ...gauntlet,
          currentStage: nextStageIndex,
          stageResults: newStageResults,
          stageStartTime: Date.now(),
          stageStartElapsedMs: action.elapsedMs ?? gauntlet.stageStartElapsedMs
        }
      };
    }

    case 'STEAL_GUESS': {
      // Opponent completed a stage ahead — reduce maxGuesses by 1 on all PLAYING boards
      if (state.mode !== GameMode.GAUNTLET || !state.gauntlet) return state;
      if (state.status !== GameStatus.PLAYING) return state;

      const stolenBoards = state.boards.map(b => {
        if (b.status !== GameStatus.PLAYING) return b;
        const newMax = Math.max(b.guesses.length + 1, b.maxGuesses - 1);
        return { ...b, maxGuesses: newMax };
      });

      return { ...state, boards: stolenBoards };
    }

    case 'BLACKOUT_RESTART': {
      // After Letter Blackout timer expires: reset the failed board with a new word
      if (state.mode !== GameMode.GAUNTLET || !state.gauntlet) return state;
      if (state.status !== GameStatus.PLAYING) return state;

      const { boardIndex } = action;
      const failedBoard = state.boards[boardIndex];
      if (!failedBoard || failedBoard.status !== GameStatus.LOST) return state;

      const gauntlet = state.gauntlet;
      const blackoutNum = gauntlet.blackoutCount + 1;
      const replacementSeed = `${state.seed}-blackout-${blackoutNum}`;
      const [newSolution] = generateSolutionsFromSeed(replacementSeed, 1);

      const stageConfig = gauntlet.stages[gauntlet.currentStage];
      const newBoard = createBoardState(newSolution, stageConfig.maxGuesses);

      if (stageConfig.hasPrefill) {
        const allowedWords = getAllowedWords();
        const prefillWords = generatePrefillWords(replacementSeed, [newSolution], allowedWords);
        newBoard.prefilledGuesses = generatePrefillGuesses(prefillWords, newSolution);
      }

      const newBoards = [...state.boards];
      newBoards[boardIndex] = newBoard;

      return {
        ...state,
        boards: newBoards,
        gauntlet: { ...gauntlet, blackoutCount: blackoutNum }
      };
    }

    case 'ABANDON': {
      return {
        ...state,
        status: GameStatus.ABANDONED
      };
    }

    case 'RESET': {
      return createInitialState(action.seed, action.mode);
    }

    case 'RESTORE_STATE': {
      // Replace entire state wholesale with a previously saved snapshot.
      // Used by persistence layers to resume mid-game after a remount.
      return action.state;
    }

    default:
      return state;
  }
}

File: apps/ios/Sources/Core/Reducer.swift
import Foundation

func initializeGame(seed: String, mode: GameMode) -> GameState {
    return createInitialState(seed: seed, mode: mode)
}

private func createBoardState(solution: String, maxGuesses: Int = 6) -> BoardState {
    return BoardState(
        solution: solution,
        guesses: [],
        maxGuesses: maxGuesses,
        status: .playing
    )
}

private func createStageBoardsFromSolutions(seed: String, stage: GauntletStageConfig, solutions: [String]) -> [BoardState] {
    let dict = GameDictionary.shared
    var boards = solutions.map { createBoardState(solution: $0, maxGuesses: stage.maxGuesses) }

    if stage.hasPrefill {
        let allowedWords = dict.getAllowedWords()
        let prefillWords = generatePrefillWords(seed: seed, solutions: solutions, allowedWords: allowedWords)
        for i in 0..<boards.count {
            boards[i].prefilledGuesses = generatePrefillGuesses(words: prefillWords, solution: boards[i].solution)
        }
    }

    return boards
}

private func getStageSolutionSlice(allSolutions: [String], stageIndex: Int) -> [String] {
    var offset = 0
    for i in 0..<stageIndex {
        offset += gauntletStages[i].boardCount
    }
    return Array(allSolutions[offset..<(offset + gauntletStages[stageIndex].boardCount)])
}

func createInitialState(seed: String, mode: GameMode) -> GameState {
    let now = Date().timeIntervalSince1970 * 1000

    switch mode {
    case .duel:
        let solutions = generateSolutionsFromSeed(seed, count: 1)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 6) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .multiDuel:
        let solutions = generateSolutionsFromSeed(seed, count: 2)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 6) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .gauntlet:
        let allSolutions = generateSolutionsFromSeed(seed, count: gauntletTotalSolutions)
        let firstStage = gauntletStages[0]
        let firstStageSolutions = getStageSolutionSlice(allSolutions: allSolutions, stageIndex: 0)
        let gauntletBoards = createStageBoardsFromSolutions(seed: seed, stage: firstStage, solutions: firstStageSolutions)

        return GameState(
            mode: mode, seed: seed, startTime: now,
            boards: gauntletBoards, currentBoardIndex: 0, status: .playing,
            gauntlet: GauntletProgress(
                currentStage: 0,
                totalStages: gauntletStages.count,
                stages: gauntletStages,
                stageResults: [],
                stageStartTime: now,
                stageStartElapsedMs: 0,
                allSolutions: allSolutions,
                blackoutCount: 0
            )
        )

    case .quordle:
        let solutions = generateSolutionsFromSeed(seed, count: 4)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 9) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .octordle:
        let solutions = generateSolutionsFromSeed(seed, count: 8)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 13) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .sequence:
        let solutions = generateSolutionsFromSeed(seed, count: 4)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 10) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .rescue:
        let solutions = generateSolutionsFromSeed(seed, count: 4)
        var boards = solutions.map { createBoardState(solution: $0, maxGuesses: 6) }
        let dict = GameDictionary.shared
        let allowedWords = dict.getAllowedWords()
        let prefillWords = generatePrefillWords(seed: seed, solutions: solutions, allowedWords: allowedWords)
        for i in 0..<boards.count {
            boards[i].prefilledGuesses = generatePrefillGuesses(words: prefillWords, solution: boards[i].solution)
        }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .tournament:
        let solutions = generateSolutionsFromSeed(seed, count: 5)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 6) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .propernoundle:
        let solutions = generateSolutionsFromSeed(seed, count: 1)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 6) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .duel6:
        let solutions = generateSolutionsFromSeedForLength(seed, count: 1, wordLength: 6)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 7) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)

    case .duel7:
        let solutions = generateSolutionsFromSeedForLength(seed, count: 1, wordLength: 7)
        let boards = solutions.map { createBoardState(solution: $0, maxGuesses: 8) }
        return GameState(mode: mode, seed: seed, startTime: now, boards: boards, currentBoardIndex: 0, status: .playing)
    }
}

func gameReducer(state: GameState, action: GameAction) -> GameState {
    switch action {

    case let .submitGuess(guess, boardIndexOpt, applyToAll):
        let dict = GameDictionary.shared
        guard dict.isValidWord(guess) else { return state }

        let updateBoard: (BoardState) -> BoardState = { board in
            guard board.status == .playing else { return board }
            guard guess.count == board.solution.count else { return board }
            let result = evaluateGuess(solution: board.solution, guess: guess)
            var newGuesses = board.guesses
            newGuesses.append(guess.uppercased())
            var newStatus = board.status
            if result.isCorrect {
                newStatus = .won
            } else if newGuesses.count >= board.maxGuesses {
                newStatus = .lost
            }
            var updated = board
            updated.guesses = newGuesses
            updated.status = newStatus
            return updated
        }

        var newBoards: [BoardState]
        if applyToAll {
            newBoards = state.boards.map(updateBoard)
            if zip(newBoards, state.boards).allSatisfy({ $0.solution == $1.solution && $0.guesses == $1.guesses }) {
                return state
            }
        } else {
            let boardIndex = boardIndexOpt ?? state.currentBoardIndex
            let board = state.boards[boardIndex]
            guard board.status == .playing else { return state }
            guard guess.count == board.solution.count else { return state }
            newBoards = state.boards
            newBoards[boardIndex] = updateBoard(board)
        }

        var gameStatus = state.status

        switch state.mode {
        case .duel, .duel6, .duel7:
            gameStatus = newBoards[0].status

        case .multiDuel:
            let allComplete = newBoards.allSatisfy { $0.status != .playing }
            if allComplete {
                let anyWon = newBoards.contains { $0.status == .won }
                gameStatus = anyWon ? .won : .lost
            }

        case .gauntlet:
            if let gauntlet = state.gauntlet {
                let anyLost = newBoards.contains { $0.status == .lost }
                let allWon = newBoards.allSatisfy { $0.status == .won }
                if anyLost && !allWon {
                    let stageGuesses = newBoards.map(\.guesses.count).max() ?? 0
                    let stageTimeMs = Int(Date().timeIntervalSince1970 * 1000 - gauntlet.stageStartTime)
                    let failedStageResult = GauntletStageResult(
                        stageIndex: gauntlet.currentStage,
                        status: .lost,
                        guesses: stageGuesses,
                        timeMs: stageTimeMs,
                        boardsSnapshot: newBoards
                    )
                    var newGauntlet = gauntlet
                    newGauntlet.stageResults.append(failedStageResult)
                    var newState = state
                    newState.boards = newBoards
                    newState.status = .lost
                    newState.gauntlet = newGauntlet
                    return newState
                }
            }

        case .quordle, .octordle, .sequence:
            let allComplete = newBoards.allSatisfy { $0.status != .playing }
            if allComplete {
                let allWon = newBoards.allSatisfy { $0.status == .won }
                gameStatus = allWon ? .won : .lost
            }

        case .rescue, .tournament:
            let allWon = newBoards.allSatisfy { $0.status == .won }
            let anyLost = newBoards.contains { $0.status == .lost }
            if allWon {
                gameStatus = .won
            } else if anyLost {
                gameStatus = .lost
            }

        case .propernoundle:
            gameStatus = newBoards[0].status
        }

        var newState = state
        newState.boards = newBoards
        newState.status = gameStatus
        return newState

    case let .submitHint(hintWord, hintEvaluation, boardIndexOpt):
        let boardIndex = boardIndexOpt ?? state.currentBoardIndex
        let board = state.boards[boardIndex]
        guard board.status == .playing else { return state }

        var newGuesses = board.guesses
        newGuesses.append(hintWord)
        let hintIndex = String(newGuesses.count - 1)

        var newStatus = board.status
        if newGuesses.count >= board.maxGuesses {
            newStatus = .lost
        }

        var updatedBoard = board
        updatedBoard.guesses = newGuesses
        updatedBoard.status = newStatus
        var hintEvals = updatedBoard.hintEvaluations ?? [:]
        hintEvals[hintIndex] = hintEvaluation
        updatedBoard.hintEvaluations = hintEvals

        var newState = state
        newState.boards[boardIndex] = updatedBoard

        switch state.mode {
        case .duel, .duel6, .duel7:
            newState.status = newStatus
        default:
            break
        }

        return newState

    case .nextBoard:
        let currentBoard = state.boards[state.currentBoardIndex]
        guard currentBoard.status == .won else { return state }
        let nextIndex = state.currentBoardIndex + 1
        guard nextIndex < state.boards.count else { return state }
        var newState = state
        newState.currentBoardIndex = nextIndex
        return newState

    case let .nextStage(elapsedMs):
        guard state.mode == .gauntlet, var gauntlet = state.gauntlet else { return state }

        let stageComplete = state.boards.allSatisfy { $0.status == .won }
        guard stageComplete else { return state }

        let stageTimeMs: Double
        if let elapsed = elapsedMs, let startElapsed = gauntlet.stageStartElapsedMs {
            stageTimeMs = max(0, elapsed - startElapsed)
        } else {
            stageTimeMs = Date().timeIntervalSince1970 * 1000 - gauntlet.stageStartTime
        }

        let stageGuesses = state.boards.map(\.guesses.count).max() ?? 0
        let stageResult = GauntletStageResult(
            stageIndex: gauntlet.currentStage,
            status: .won,
            guesses: stageGuesses,
            timeMs: Int(stageTimeMs),
            boardsSnapshot: state.boards
        )

        gauntlet.stageResults.append(stageResult)
        let nextStageIndex = gauntlet.currentStage + 1

        if nextStageIndex >= gauntlet.totalStages {
            var newState = state
            newState.status = .won
            newState.gauntlet = gauntlet
            return newState
        }

        let nextStage = gauntlet.stages[nextStageIndex]
        let nextSolutions = getStageSolutionSlice(allSolutions: gauntlet.allSolutions, stageIndex: nextStageIndex)
        let nextBoards = createStageBoardsFromSolutions(seed: state.seed, stage: nextStage, solutions: nextSolutions)
        let now = Date().timeIntervalSince1970 * 1000

        gauntlet.currentStage = nextStageIndex
        gauntlet.stageStartTime = now
        gauntlet.stageStartElapsedMs = elapsedMs ?? gauntlet.stageStartElapsedMs

        var newState = state
        newState.boards = nextBoards
        newState.currentBoardIndex = 0
        newState.gauntlet = gauntlet
        return newState

    case .stealGuess:
        guard state.mode == .gauntlet, state.gauntlet != nil else { return state }
        guard state.status == .playing else { return state }

        var newState = state
        for i in 0..<newState.boards.count {
            guard newState.boards[i].status == .playing else { continue }
            let newMax = max(newState.boards[i].guesses.count + 1, newState.boards[i].maxGuesses - 1)
            newState.boards[i].maxGuesses = newMax
        }
        return newState

    case let .blackoutRestart(boardIndex):
        guard state.mode == .gauntlet, var gauntlet = state.gauntlet else { return state }
        guard state.status == .playing else { return state }
        guard boardIndex < state.boards.count, state.boards[boardIndex].status == .lost else { return state }

        let blackoutNum = gauntlet.blackoutCount + 1
        let replacementSeed = "\(state.seed)-blackout-\(blackoutNum)"
        let newSolutions = generateSolutionsFromSeed(replacementSeed, count: 1)
        let newSolution = newSolutions[0]

        let stageConfig = gauntlet.stages[gauntlet.currentStage]
        var newBoard = createBoardState(solution: newSolution, maxGuesses: stageConfig.maxGuesses)

        if stageConfig.hasPrefill {
            let dict = GameDictionary.shared
            let allowedWords = dict.getAllowedWords()
            let prefillWords = generatePrefillWords(seed: replacementSeed, solutions: [newSolution], allowedWords: allowedWords)
            newBoard.prefilledGuesses = generatePrefillGuesses(words: prefillWords, solution: newSolution)
        }

        gauntlet.blackoutCount = blackoutNum
        var newState = state
        newState.boards[boardIndex] = newBoard
        newState.gauntlet = gauntlet
        return newState

    case .abandon:
        var newState = state
        newState.status = .abandoned
        return newState

    case let .reset(seed, mode):
        return createInitialState(seed: seed, mode: mode)

    case let .restoreState(restoredState):
        return restoredState
    }
}
